File: Python/14888_연산자끼워넣기.py
def calNum(N,A,s,oper,now,r_max,r_min):
	if s == N: # 최대값 최소값 갱신
		r_max = max(now,r_max)
		r_min = min(now,r_min)
		
		return r_max, r_min
		
	for i in range(4):
		if oper[i] > 0: # 사용가능한 연산자 1개일 때
			oper[i] -= 1 # 현재 사용할 연산자의 수 1감소
			if i == 0: # 덧셈
				r_max, r_min = calNum(N,A,s+1,oper,now+A[s],r_max,r_min)
			elif i == 1: # 뺄셈
				r_max, r_min = calNum(N,A,s+1,oper,now-A[s],r_max,r_min)
			elif i == 2: # 곱셈
				r_max, r_min = calNum(N,A,s+1,oper,now*A[s],r_max,r_min)
			else: # 나눗셈
				r_max, r_min = calNum(N,A,s+1,oper,int(now/A[s]),r_max,r_min)
			oper[i] += 1 # 백트래킹 위해 원상복구
			
	return r_max, r_min

if __name__ == '__main__':
	N = int(input())
	A = list(map(int,input().split()))
	oper = list(map(int,input().split()))
	
	r_max, r_min = calNum(N,A,1,oper,A[0],float('-inf'),float('inf'))
	
	print(r_max)
	print(r_min)
	

File: C++/14888_연산자끼워넣기.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int> calNum(int N, int s, int A[], vector<int> oper, vector<int> result, int now, bool visit[]);

int main() {
	int N;
	cin >> N;
	
	int A[N];
	for(int i = 0; i < N; i++){
		cin >> A[i];
	}
	
	vector<int> oper; // 연산자 담기
	for(int i = 0; i < 4; i++){
		int tmp;
		cin >> tmp;
		for(int j = 0; j < tmp; j++){
			oper.push_back(i);
		}
	}
	
	bool visit[N-1] = {0,}; // 연산자 사용 여부
	vector<int> result(2); // 최대 최소를 담을 vector
	result[0] = -1000000001; //최대값을 구하기 위한 초기값
	result[1] = 1000000001; // 최소값을 구하기 위한 초기값
	result = calNum(N,1,A,oper,result,A[0],visit);
	cout << result[0] << '\n' << result[1];
	
	return 0;
}

vector<int> calNum(int N, int s, int A[], vector<int> oper, vector<int> result, int now, bool visit[]){
	if(s == N){
		if(now > result[0]){
			result[0] = now; // 최대값 갱신
		}
		if(now < result[1]){ // 최소값 갱신
			result[1] = now;
		}
		
		return result;
	}
	
	for(int i = 0; i < N-1; i++){
		if(visit[i] == false){ // 사용한 적 없는 연산자라면
			visit[i] = true; // 연산자 사용 표시
			
			switch(oper[i]){
				case 0: // 더하기
					result = calNum(N,s+1,A,oper,result,now+A[s],visit);
					break;
				case 1: // 빼기
					result = calNum(N,s+1,A,oper,result,now-A[s],visit);
					break;
				case 2: // 곱하기
					result = calNum(N,s+1,A,oper,result,now*A[s],visit);
					break;
				case 3: // 나누기
					result = calNum(N,s+1,A,oper,result,now/A[s],visit);
					break;
			}
			
			visit[i] = false; // 연산자 사용 취소(백트래킹위해)
		}
	}
	
	return result;
}
